### DISCUSSION TITLE
Non-blocking communication with engine

### DISCUSSION THREAD

**Discussion by clarke-baxter** [niklasf/python-chess#1061] [Q&A] [answered]
Like many others I have built a reed switch chessboard (this one is big!) It has a Teensy 4.1 to do the scanning and move validation.  The board accepts a FEN string and if the position matches the FEN it will allow a move to be made. I did it this way so that sliding moves, knocked over pieces, clumsy captures etc. all work. Pieces can dance around the board on their way to a destination if they like. If the move is legal it will be sent via serial to a Raspberry Pi 4b. On the Pi I have built a touchscreen GUI using PySImpleGUI. I am using python-chess (awesome) for managing the games, pgn files and FEN generation for the board. 

I have reached a stage where human v. human games can be played, recorded, e-mailed etc. The GUI is responsive during game play and I have smooth running clocks.

Now it is time to interact with Stockfish / UCI engines.

I can launch an engine and interact with it. The problem is that whilst waiting for a response from the engine my GUI / clocks stop. This is not unexpected - it was always going to hold up my main loop - that's where the code is to wait on the response.

So I need to be able to send a command to the engine and then in my main loop check to see if there is a response. It is fair to say I am a Python 'newbie' and I'm struggling to understand where to start. Do I need threading, mutltprocessing, asyncio? Many people must have exactly this thing working.

I have looked at the asyncio chess.engine examples, but I cannot get my head around the flow of execution and how it might be integrated into my code.

Any help much appreciated.

**Comment by niklasf** [comment, marked as answer]
`asyncio` would have the lowest overhead, but it looks like PySimpleGUI does not provide integration.

The next best thing is to use the blocking `chess.engine.SimpleEngine`. I found this example for how to deal with blocking operations in PySimpleGUI: https://github.com/PySimpleGUI/PySimpleGUI/blob/master/DemoPrograms/Demo_Multithreaded_Long_Tasks.py. So `window.start_thread()` can be used to start work that would otherwise block the main thread, and the thread will be able to send events back to the main thread.

`chess.engine.SimpleEngine` is thread-safe as long as objects passed to methods are not concurrently modified. So you could create one instance and use it in the background threads.

**Reply by clarke-baxter** [reply]
Brilliant! I was fixated on Python... it did not occur to me to look for an answer in PySimpleGUI.
Thank you for bailing me out again.

Your python-chess has enabled me to make speech and GUI add-on extensions for 1980s chess computers, and now also my own unique build. Thank you for that too.

**Comment by JohanSjoblom** [comment]
One year later and I noticed a little bit the same issue.
I use the async version and call await play(). That actually blocks ny main async loop. I had to use 2 things to avoid blocking 
1) create a new async task and run the play in that task.
2) not wait for the result of play outside of the created task. Instead I had to add an async queue. The created task awaits blocked by play() and when ready writes to the queue. The main async loop only awaits for a queue item.

These 2 steps above allows me to call play() without being blocked.

PS, I use this solution for a DGT board. In
https://github.com/JohanSjoblom/picochess

Best regards, Johan
